nfsd.md

```diff
@@ -0,0 +1,287 @@
+# NFS Server Subsystem Delta
+
+## Overview
+
+NFSD (fs/nfsd/) implements the Linux NFS server with support for NFSv2, NFSv3,
+and NFSv4.x protocols. The subsystem is actively migrating NFSv3 XDR handling
+to auto-generated code (xdrgen). Code is generally well-structured but complex
+state management and reference counting require careful review.
+
+## File Risk Classification
+
+Use this to prioritize review depth:
+
+**CRITICAL (>9000 lines, complex state):**
+- nfs4state.c (9371 lines) - Client/session/lock/delegation management
+  - Watch: refcount leaks, locking order, state machine bugs
+
+**HIGH RISK (3000-6000 lines):**
+- nfs4xdr.c (6034 lines) - Manual NFSv4 XDR encoding/decoding
+  - Watch: bounds checking, buffer overflow
+- nfs4proc.c (3921 lines) - NFSv4 procedure handlers
+  - Watch: state transitions, error handling
+
+**MEDIUM RISK (1000-3000 lines):**
+- vfs.c (2914 lines) - I/O hot paths, VFS interface
+  - Watch: bounds in loops, performance-critical code
+- export.c (1547 lines) - Access control (security-critical)
+  - Watch: input validation, permission checks
+- nfsfh.c (893 lines) - File handle validation (security-critical)
+  - Watch: reference counting, error paths
+- nfs3proc.c (1321 lines) - NFSv3 handlers (being migrated)
+  - Watch: error code mapping, cleanup
+
+**LOW RISK (auto-generated):**
+- nfs3xdr_gen.c (3768 lines) - Generated from formal spec
+  - Review: verify struct sizes, procedure linkage only
+
+## NFSD-Specific Patterns [NFSD]
+
+#### NFSD-001: XDR bounds validation
+
+**Risk**: Buffer overflow from untrusted network input
+
+**Check:**
+- All string lengths validated before use
+- Array counts checked against reasonable limits
+- Loop iterations in hot paths have upper bounds
+- xdr_reserve_space_vec() called BEFORE encoding
+
+**Example vulnerability:** nfsd_iter_read() missing max_read check (7ac3be9e56d8)
+
+#### NFSD-002: Reference counting balance
+
+**Risk**: Use-after-free, double-free, resource leak
+
+**Check:**
+- Every fh_copy() has matching fh_put()
+- Every exp_get() has matching exp_put()
+- Every atomic_inc() balanced by atomic_dec()
+- Every kref_get() balanced by kref_put()
+- Resources assigned AFTER all error checks complete
+- Error paths properly decrement refcounts
+
+**Example vulnerability:** Early assignment in nfsd_set_fh_dentry() (b3da9b141578)
```

Comment:
This seems really useful in general.  Can you please add it to patterns/RM-003.md?